### PR TITLE
GH-40855: [C++][ORC] Fix `std::filesystem` related link error with ORC 2.0.0 or later

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

If we use `std::filesystem`, we need `-lstdc++fs` with GCC 8 and `-lc++fs` for clang 7. We don't want to maintain CMake code for GCC 8/clang 7.

### What changes are included in this PR?

* Don't use `std::filesystem` with ORC 2.0.0 or later.
* Use missing `-lstdc++fs`/`-lc++fs` with bundled ORC 2.0.0.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.
* GitHub Issue: #40855

### CODE REVIEW

**Comment by kou** [comment]
@github-actions crossbow submit almalinux-8-amd64

**Comment by kou** [comment]
My guess was wrong...
ORC 2.0.0 or later uses `std::filesystem`. So we need to specify `-lstdc++fs` or `-lc++fs` to use `liborc.a`.

File: cpp/cmake_modules/FindorcAlt.cmake [hunk 1]
```diff
@@ -29,6 +29,7 @@ endif()
 find_package(orc ${find_package_args})
 if(orc_FOUND)
   set(orcAlt_FOUND TRUE)
+  set(orcAlt_VERSION ${orc_VERSION})
```
Comment (wgtmac) [inline cpp/cmake_modules/FindorcAlt.cmake:32]: Does this work with conan？

**Comment by wgtmac** [comment]
> My guess was wrong... ORC 2.0.0 or later uses `std::filesystem`. So we need to specify `-lstdc++fs` or `-lc++fs` to use `liborc.a`.

Is any other module doing the same thing, or ORC is the only one that requires std::filesystem?

**Comment by kou** [comment]
@github-actions crossbow submit almalinux-8-amd64

File: cpp/cmake_modules/FindorcAlt.cmake (hunk 1 above)
Comment (kou) [reply cpp/cmake_modules/FindorcAlt.cmake:32]: I hope so...
Do you know conan's ORC recipe provide valid(?) CMake package for ORC?

(BTW, is there a plan that ORC provides its CMake package?)

**Comment by kou** [comment]
> Is any other module doing the same thing, or ORC is the only one that requires std::filesystem?

The latter.

File: cpp/cmake_modules/FindorcAlt.cmake (hunk 1 above)
Comment (wgtmac) [reply cpp/cmake_modules/FindorcAlt.cmake:32]: > Do you know conan's ORC recipe provide valid(?) CMake package for ORC?

I suppose it does. It was added by me: https://github.com/conan-io/conan-center-index/blob/master/recipes/orc/all/conanfile.py#L145

> is there a plan that ORC provides its CMake package?

I'm not sure if I understand the question correctly, do you mean the ORC community provides an official FindOrc.cmake file or something similar?

**Comment by wgtmac** [comment]
> > Is any other module doing the same thing, or ORC is the only one that requires std::filesystem?
> 
> The latter.

I can fix that on the ORC side by removing std::filesystem dependency in 2.0.1 if you think appropriate and revert ORC 2.0.0 to 1.9.3.

**Comment by kou** [comment]
> I can fix that on the ORC side by removing std::filesystem dependency in 2.0.1 if you think appropriate and revert ORC 2.0.0 to 1.9.3.

It'll be helpful. Could you do it? But we don't need to revert bundled ORC to 1.9.3. Time zone DB path check in ORC 2.0.0 will be useful. And this PR will fix this `std::filesystem` problem.

**Comment by wgtmac** [comment]
> > I can fix that on the ORC side by removing std::filesystem dependency in 2.0.1 if you think appropriate and revert ORC 2.0.0 to 1.9.3.
> 
> It'll be helpful. Could you do it? But we don't need to revert bundled ORC to 1.9.3. Time zone DB path check in ORC 2.0.0 will be useful. And this PR will fix this `std::filesystem` problem.

OK, I will remove std::filesystem dependency but keep the TZDB check on the ORC side.

File: cpp/cmake_modules/FindorcAlt.cmake (hunk 1 above)
Comment (wgtmac) [reply cpp/cmake_modules/FindorcAlt.cmake:32]: OK, let me investigate how does it work and add it to ORC.

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp

**Review by wgtmac — APPROVED** [review]

**Pull request merged**